### PR TITLE
Sockets - Grace

### PR DESCRIPTION
**Pull request by gracemshea**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I have a model that returns the works with the top 10 by votes by category.
Describe how you approached testing that model method. What edge cases did you come up with? |  When testing in general I generated 10 works and test for them to be present. For edges, I tested that it would return the entire list of work if there were less than 10 or an empty array for no works.
What are `session` and `flash`? What is the difference between them? | Session and flash are hash-like objects that persist through request-response cycle. Session lasts until the user exits the browser (or logs out which essentially sets the user of the session to nil). Flash persists over one cycle and is cleared after the next action.
What was one thing that you gained more clarity on through this assignment? | Validations, Relationships, and Custom Methods in models and model testing.
What is the Heroku URL of your deployed application? | https://grace-media-ranker-app.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I felt really disorganized on this project. I think it was related to learning things in class and feeling like I was starting aspects of my project over to try and apply that knowledge. A mid-week check in might have been nice, although I'm not sure if others were quite as scattered/challenged as me.

### CODE REVIEW

File: app/views/users/login_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<h1>User Log In</h1>
+
+<% form_with model: @user, url: login_path do |f| %>
+  <div class="form-group">
+    <%= f.label :name %>
+    <%= f.text_field :name, placeholder: "Username", class:"form-control" %>
```
Comment (droberts-sea) [inline app/views/users/login_form.html.erb:6]: Line 3 should be a value tag (`<%=`), otherwise the form won't appear on the page.

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class Work < ApplicationRecord
+  has_many :votes
+  has_many :users, through: :votes, dependent: :destroy
+  validates :title, presence: true
+
+  def self.get_media_categories
+    books = self.get_media("books")
+    albums = self.get_media("albums")
+    movies = self.get_media("movies")
```
Comment (droberts-sea) [inline app/models/work.rb:9]: Here you search for works with category `albums`, `books` and `movies`, but your database seeds and the dropdown menu on your new work form use the singular version of these (`album`, `book`, `movie`). As a result, your index and splash pages don't display any works.

File: app/views/works/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      <%= @work.description %><br>
+    </section>
+</header>
+  
+  <section class="actions">
+    <%= link_to "Back to media ranks", root_path, class: "btn btn-primary btn-lg" %>
+    <%= link_to "Edit", edit_work_path(@work.id), class: "btn btn-primary btn-lg" %>
+    <%= link_to "Upvote", class: "btn btn-primary btn-lg" %>
+    <%= link_to "Delete", class: "btn btn-danger btn-lg" %> 
+  </section>
```
Comment (droberts-sea) [inline app/views/works/show.html.erb:15]: The links on line 13 and 14 are missing URLs, so neither of them works.

File: app/views/homepages/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+
+<section class="spotlight">
+<% media = Work.all.sample%>
+    <h2 class="spotlight__header">
+        <span class="spotlight__header--prefix"> Media Spotlight </span>
```
Comment (droberts-sea) [inline app/views/homepages/index.html.erb:5]: This should probably use the `@spotlight` variable that you set up in the controller. As-is it shows a random work, not the one with the most votes.

File: app/views/homepages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        <a class="spotlight__description"><%= link_to media.title, work_path(media.id)%> by <%= media.creator%></a>
+    </h2>
+    <p class="spotlight__description"><%= media.description%></p>
+</section>
+
+<section class="top-ten__container">
+  <section class="top-ten__list-container"> 
+    <h2>Top Movies</h2>
+      <%= render "top_ten", top: @top_ten[2] %>
+  </section>
```
Comment (droberts-sea) [inline app/views/homepages/index.html.erb:15]: I love the way you've used a partial to DRY this up - it makes this code much more readable.

File: app/views/works/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ <thead>
+  <th>User</th>
+  <th>Date</th>
+  </thead>
+  <tbody>
+    <% @work.votes.each do |vote| %>
+    <tr>
+      <td><%= link_to vote.user, user_path(vote.user) %></td>
+      <td> <%= vote.date %> </td>
+    </tr>
```
Comment (droberts-sea) [inline app/views/works/show.html.erb:30]: This table doesn't quite render the way you probably intended. Something like

```html+erb
<tr>
  <td><%= link_to vote.user.name, user_path(vote.user) %></td>
  <td><%= vote.created_at %></td>
</tr>
```

might be closer to what you want.

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root "homepages#index"
+  get "/", to: "homepages#index"
+  resources :works, only: [:index, :show, :new, :create, :edit, :update, :destroy]
+  resources :works do
+    resources :votes, only: [:create]
```
Comment (droberts-sea) [inline config/routes.rb:8]: This code is a little redundant:

- On line 6, you limit `resources` to generate "only" all 7 of the RESTful routes
- On line 7, you re-generate all 7 routes again

The change here would be to omit line 6.

File: config/routes.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  post "/logout", to: "users#logout", as: "logout"
+  get "/login", to: "users#login_form", as: "login"
+  post "/login", to: "users#login"
+  get "/users/current", to: "users#current", as: "current_user"
+
+  resources :users, only: [:index, :show]
```
Comment (droberts-sea) [inline config/routes.rb:16]: Good work generating only the routes your application needs.

File: app/models/vote.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+  validates :user, uniqueness: { scope: [:work], message: "has already voted for this work" }
```
Comment (droberts-sea) [inline app/models/vote.rb:4]: Good work getting this tricky `uniqueness` scope figured out!

File: app/models/work.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  def self.get_media(media_type)
+    media = Work.where(category: media_type)
+  end
+
+  # return top ten works
+  def self.top_media
+    top_works = get_media_categories.map do |category|
+      category[0..9]
+    end
```
Comment (droberts-sea) [inline app/models/work.rb:21]: I love the way the methods in this file build on each other, each providing a small bit of functionality. This is an excellent example of functional decomposition.

File: test/controllers/users_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    end
+    it "will login a returning user" do
+      post login_path(@login_data)
+
+      expect(session[:user_id]).must_equal @user.id
+    end
+
+    it "will create an account for and login a new user" do
+      @login_data[:user][:name] = "new user"
+
```
Comment (droberts-sea) [inline test/controllers/users_controller_test.rb:68]: Good work catching the two different success cases here.

Is there a failure case? What if the user forgets to type in their name before logging in?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |
Comprehension questions |
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes - good work!
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | yes
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | yes - great work!
**Wave 1 - Media** |
Splash page shows the three media categories | Almost - you have a bug in your `Work.get_media_categories` method
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | no
Splash page contains a media spotlight | yes, but see inline
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes, but see inline
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** |


This is a strong submission. Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met.


However, there are a number of bugs in this app that were small and straightforward to fix, but which had a devastating effect on the user experience, like not showing any works on the splash or index pages, not displaying the login form, or not being able to upvote or delete from the show page.



What that tells me is that you're relying exclusively on your tests to tell you whether your site works. Your testing on this assignment is really strong, but there were a couple of places where it didn't quite match up with what the user is doing. These are all things that manually running through the workflows of your site would have caught immediately.


**The takeaway is, before you submit your next assignment, spend 10 minutes running through it manually and making sure it works!**
